### PR TITLE
Update Truffle

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

**Comment by chrisseaton** [comment]
We need Sulong to update first.

File: src/main/java/org/truffleruby/RubyLanguage.java [hunk 1]
```diff
@@ -107,6 +107,7 @@ protected CallTarget parse(ParsingRequest request) throws Exception {
         return Truffle.getRuntime().createCallTarget(new LazyRubyRootNode(this, null, null, request.getSource(), request.getArgumentNames()));
     }
 
+    @SuppressWarnings("deprecation")
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/RubyLanguage.java]: Do we have a way to audit these? Should you add a TODO here?

File: src/main/java/org/truffleruby/interop/InteropNodes.java [hunk 2]
```diff
@@ -939,11 +939,25 @@ public boolean internalBit(int bits) {
...
+            if (readable) {
+                keyInfo |= KeyInfo.READABLE;
+            }
+
+            if (writable) {
+                keyInfo |= KeyInfo.MODIFIABLE;
+            }
+
+            if (invocable) {
+                keyInfo |= KeyInfo.INVOCABLE;
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/interop/InteropNodes.java]: Just a casual observation, but I was surprised to see this wasn't name `INVOKABLE`.

File: spec/truffle/interop/key_info_spec.rb [hunk 3]
```diff
@@ -10,6 +10,90 @@
 require_relative 'fixtures/classes'
 
 describe "Truffle::Interop.key_info" do
+  describe "for an Array" do
+    
+    before :each do
+      @array = [1, 2, 3]
+    end
+    
+    it "returns :readable for indicies in bounds" do
```
Comment (nirvdrum) [inline spec/truffle/interop/key_info_spec.rb]: "indicies" -> "indices"? (Throughout this file if you agree).

File: src/main/java/org/truffleruby/RubyLanguage.java (hunk 1 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/RubyLanguage.java]: I'll open an issue for them.

File: spec/truffle/interop/key_info_spec.rb (hunk 3 above)
Comment (chrisseaton) [reply spec/truffle/interop/key_info_spec.rb]: Fixed.

**Comment by chrisseaton** [comment]
I think duck-typing is ok there - if you try to hash it and it fails you'll get a reasonably clear error message and that's usual for Ruby.

File: doc/user/interop.md [hunk 4]
```diff
@@ -113,42 +113,46 @@ If the object is a Ruby `Array` and the key is an integer:
...
 
 `READABLE` will be set if the object responds to a method of the same name.
 
-`WRITABLE` will be set if the object responds to a method of the same name
+`MODIFIABLE` will be set if the object responds to a method of the same name
 appended with `=` and the object is not frozen.
 
+`INSERTABLE` will be false.
+
 If they key has a leading `@`:
```
Comment (eregon) [inline doc/user/interop.md]: If `the` key

File: spec/truffle/interop/key_info_spec.rb [hunk 5]
```diff
@@ -10,6 +10,40 @@
...
+      end
+    end
+    
+    it "returns :removable for indicies out of bounds" do
+      [-1, 3].each do |n|
+        Truffle::Interop.key_info(@array, n).should include(:removable)
+      end
+    end
+    
+    it "returns :removable for indicies out of bounds if the array is frozen" do
```
Comment (eregon) [inline spec/truffle/interop/key_info_spec.rb]: does not ...

File: spec/truffle/interop/key_info_spec.rb [hunk 6]
```diff
@@ -54,6 +110,12 @@
       end
     end
     
+    it "returns :removable for all keys" do
+      @hash.keys.each do |k|
+        Truffle::Interop.key_info(@hash, k).should include(:removable)
+      end
+    end
```
Comment (eregon) [inline spec/truffle/interop/key_info_spec.rb]: Isn't this duplicated with a spec above?

File: doc/user/interop.md (hunk 4 above)
Comment (chrisseaton) [reply doc/user/interop.md]: Fixed in #1217 (but wasn't actually introduced in this PR anyway).

File: spec/truffle/interop/key_info_spec.rb (hunk 5 above)
Comment (chrisseaton) [reply spec/truffle/interop/key_info_spec.rb]: Fixed in #1217.

File: spec/truffle/interop/key_info_spec.rb (hunk 6 above)
Comment (chrisseaton) [reply spec/truffle/interop/key_info_spec.rb]: Fixed in #1217.

**Pull request merged**